int8 phase
int8 target_id